[css-flexbox][baseline-alignment] Use BaslineContext/BaselineGroup to perform baseline alignment.

sammygill opened the pull request:
#### b7b43b9bf044a2edb16f3fc873f69f1bdf3d8379
<pre>
[css-flexbox][baseline-alignment] Use BaslineContext/BaselineGroup to perform baseline alignment.
<a href="https://bugs.webkit.org/show_bug.cgi?id=256793">https://bugs.webkit.org/show_bug.cgi?id=256793</a>
rdar://109362699

Reviewed by Alan Baradlay.

The BaselineContext and BaselineGroup classes provide us with the ability
to represent baseline alignment contexts and baseline sharing groups
as described in css-align-3. The main purpose of this patch is to replace
the data structures we use in flex layout to hold state for baseline
alignment with these new objects.

Previously, we were storing our baseline alignment state, in particular
the max ascent for baseline and last baseline alignment, for each
flex line inside of LineContext. The maxAscent value was computed
off of all of the items which had baseline alignment specified and
lastBaselineMaxAscent was computed off of all of the items which had
last baseline alignment specified.

This was not quite right since the spec mentions ways in which items
there are baseline and last baseline aligned could get placed in the
same baseline sharing group and should get aligned together. In the
following example the two flex items are placed in the same baseline
sharing group:

&lt;div style=&quot;display: flex; flex-direction: column; align-items: last-baseline; width: 200px;&quot;&gt;
  &lt;div style=&quot;align-self: first baseline; writing-mode: vertical-rl;&quot;&gt;line1&lt;br&gt;line2&lt;/div&gt;
  &lt;div style=&quot;writing-mode: vertical-lr;&quot;&gt;line1&lt;br&gt;line2&lt;/div&gt;
&lt;/div&gt;

With the new addition of the BaselineContext and BaselineGroup objects
we can now place items into the correct baseline sharing group as
described in the spec and the max ascent values are now kept track of
within the baseline sharing group. However, the above example does
still not render correctly because there are still some outstanding
bugs related to baseline alignment, but this serves as a good starting
point for addressing them.

In order to perform baseline alignment with these new classes, a slight
refactoring was needed. Instead of iterating over all of the children
on the flex line we instead iterate over all of the baseline sharing
groups on the line and perform the alignment for each group. We do this
by checking to see if the LineContext has a BaselineContext early in
alignChildren() and for into performBaselineAlignment(). This is a
little more intuitive and will be useful as we address the rest of the
portions of the spec in css-align-3 as it mentions working with baseline
sharing groups as a whole.

Since we have not added any new functionality in regards to baseline
alignment besides just separating items into baseline sharing groups,
we should ideally not be regressing any scenarios in which we perform
baseline and last baseline alignment correctly.

* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-004-expected.txt:
* Source/WebCore/rendering/BaselineAlignment.cpp:
(WebCore::BaselineAlignmentState::sharedGroups):
* Source/WebCore/rendering/BaselineAlignment.h:
(WebCore::BaselineGroup::begin):
(WebCore::BaselineGroup::end):
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::LineState::LineState):
(WebCore::alignmentOffset):
(WebCore::RenderFlexibleBox::staticCrossAxisPositionForPositionedChild):
(WebCore::RenderFlexibleBox::layoutAndPlaceChildren):
(WebCore::RenderFlexibleBox::alignChildren):
(WebCore::RenderFlexibleBox::performBaselineAlignment):
* Source/WebCore/rendering/RenderFlexibleBox.h:

Canonical link: <a href="https://commits.webkit.org/266387@main">https://commits.webkit.org/266387@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/944b2f63a881baf5de67d545981a5aceed93a21c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/13672 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/13986 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/14319 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/15408 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/12983 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/13753 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/16494 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/14067 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/15665 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/13838 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/14462 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/11568 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/16105 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/11751 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/12326 "2 flakes 5 failures") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/19360 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/12826 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/12491 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/15698 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/13017 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/10897 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/12287 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3331 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/16618 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/12858 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/rendering/BaselineAlignment.cpp:
```diff
@@ -88,6 +88,11 @@ const BaselineGroup& BaselineContext::sharedGroup(const RenderBox& child, ItemPo
     return const_cast<BaselineContext*>(this)->findCompatibleSharedGroup(child, preference);
 }
 
+Vector<BaselineGroup> BaselineContext::sharedGroups() const
```
this seems to return a _copy_ of the vector. is that what callers expect/need here?

sammygill replied:
Nope this should return a reference. Also had to make `baselineContext` in `LineContext` mutable since the alignment code uses `const LineContext`. I think this is better than removing the `const` from the function argument so we can make sure the other members of the struct don’t get mutated, but let me know if you think we should do that instead or take another approach for that

alanbaradlay replied:
"context" to me reads like something you don't mutate but instead you change the associated "state". Maybe it's just incorrectly named structs/classes, not sure, but I think even in the commit message you referred to some of these things as "state". 
Seeing a single member of a "context" class mutable (e.g. while debugging some issues) would certainly make me scratch my head. :)

sammygill replied:
Yeah so I removed the mutable from the now `baselineState` and now whenever we need to modify it I also removed the const from the `LineState`

alanbaradlay commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -1902,7 +1900,7 @@ LayoutUnit RenderFlexibleBox::staticMainAxisPositionForPositionedChild(const Ren
 {
     auto availableSpace = availableAlignmentSpaceForChild(crossAxisContentExtent(), child);
-    return alignmentOffset(availableSpace, alignmentForChild(child), 0_lu, 0_lu, style().flexWrap() == FlexWrap::Reverse);
+    return alignmentOffset(availableSpace, alignmentForChild(child), std::nullopt, std::nullopt, style().flexWrap() == FlexWrap::Reverse);
```
I usually just go with { }, -but this is just a matter of taste.

sammygill committed:
[css-flexbox][baseline-alignment] Use BaslineContext/BaselineGroup to perform baseline alignment.
https://bugs.webkit.org/show_bug.cgi?id=256793
rdar://109362699

Reviewed by Alan Baradlay.

The BaselineContext and BaselineGroup classes provide us with the ability
to represent baseline alignment contexts and baseline sharing groups
as described in css-align-3. The main purpose of this patch is to replace
the data structures we use in flex layout to hold state for baseline
alignment with these new objects.

Previously, we were storing our baseline alignment state, in particular
the max ascent for baseline and last baseline alignment, for each
flex line inside of LineContext. The maxAscent value was computed
off of all of the items which had baseline alignment specified and
lastBaselineMaxAscent was computed off of all of the items which had
last baseline alignment specified.

This was not quite right since the spec mentions ways in which items
there are baseline and last baseline aligned could get placed in the
same baseline sharing group and should get aligned together. In the
following example the two flex items are placed in the same baseline
sharing group:

<div style="display: flex; flex-direction: column; align-items: last-baseline; width: 200px;">
  <div style="align-self: first baseline; writing-mode: vertical-rl;">line1<br>line2</div>
  <div style="writing-mode: vertical-lr;">line1<br>line2</div>
</div>

With the new addition of the BaselineContext and BaselineGroup objects
we can now place items into the correct baseline sharing group as
described in the spec and the max ascent values are now kept track of
within the baseline sharing group. However, the above example does
still not render correctly because there are still some outstanding
bugs related to baseline alignment, but this serves as a good starting
point for addressing them.

In order to perform baseline alignment with these new classes, a slight
refactoring was needed. Instead of iterating over all of the children
on the flex line we instead iterate over all of the baseline sharing
groups on the line and perform the alignment for each group. We do this
by checking to see if the LineContext has a BaselineContext early in
alignChildren() and for into performBaselineAlignment(). This is a
little more intuitive and will be useful as we address the rest of the
portions of the spec in css-align-3 as it mentions working with baseline
sharing groups as a whole.

Since we have not added any new functionality in regards to baseline
alignment besides just separating items into baseline sharing groups,
we should ideally not be regressing any scenarios in which we perform
baseline and last baseline alignment correctly.

* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-004-expected.txt:
* Source/WebCore/rendering/BaselineAlignment.cpp:
(WebCore::BaselineAlignmentState::sharedGroups):
* Source/WebCore/rendering/BaselineAlignment.h:
(WebCore::BaselineGroup::begin):
(WebCore::BaselineGroup::end):
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::LineState::LineState):
(WebCore::alignmentOffset):
(WebCore::RenderFlexibleBox::staticCrossAxisPositionForPositionedChild):
(WebCore::RenderFlexibleBox::layoutAndPlaceChildren):
(WebCore::RenderFlexibleBox::alignChildren):
(WebCore::RenderFlexibleBox::performBaselineAlignment):
* Source/WebCore/rendering/RenderFlexibleBox.h:

Canonical link: https://commits.webkit.org/266387@main
